spec/workspace-spec.coffee

```diff
@@ -941,6 +941,113 @@ describe "Workspace", ->
+              # guaranteed, so we can only verify the total number of paths searched is correct
+              # after the second call.
+              expect(onPathsSearched.callCount).toBe(2)
+              expect(onPathsSearched.mostRecentCall.args[0]).toBe(
+                numPathsToPretendToSearchInCustomDirectorySearcher + numPathsSearchedInDir2)
+
+          it "can be cancelled when the object returned by scan() has its cancel() method invoked", ->
+            lastCustomDirectorySearchCreated = null
+            class CustomDirectorySearch
+              constructor: ->
+                lastCustomDirectorySearchCreated = this
+                @promise = Promise.resolve()
+              then: (args...) ->
+                @promise.then.apply(@promise, args)
+              cancel: ->
+
+            class CustomDirectorySearcher
+              canSearchDirectory: (directory) -> directory.getPath() is dir1
+              search: (directory, delegate, options) ->
+                new CustomDirectorySearch
+
+            atom.packages.serviceHub.provide(
+              "atom.directory-searcher", "0.1.0", new CustomDirectorySearcher())
+
+            thenable = atom.workspace.scan /aaaa/, ->
+            cancelSpy = spyOn(lastCustomDirectorySearchCreated, 'cancel').andCallThrough()
+            expect(cancelSpy).not.toHaveBeenCalled()
+            thenable.cancel()
+            expect(cancelSpy).toHaveBeenCalled()
+
+            resultOfPromiseSearch = null
+            waitsForPromise ->
+              thenable.then (promiseResult) -> resultOfPromiseSearch = promiseResult
+
+            runs ->
+              expect(resultOfPromiseSearch).toBe('cancelled')
+
+          it "will have the side-effect of failing the overall search if it fails", ->
+            # Note that hoisting reject in this way is generally frowned upon.
+            hoistedReject = null
+            class CustomDirectorySearchThatWillFail
+              constructor: ->
+                @promise = new Promise (resolve, reject) ->
+                  hoistedReject = reject
+              then: (args...) ->
+                @promise.then.apply(@promise, args)
+              cancel: ->
+
+            class CustomDirectorySearcherToCancel
+              canSearchDirectory: (directory) -> directory.getPath() is dir1
+              search: (directory, options) ->
+                new CustomDirectorySearchThatWillFail
+
+            atom.packages.serviceHub.provide(
+              "atom.directory-searcher", "0.1.0", new CustomDirectorySearcherToCancel())
+
+            cancelableSearch = atom.workspace.scan /aaaa/, ->
+            hoistedReject()
+
+            resultOfPromiseSearch = null
+            waitsForPromise ->
+              cancelableSearch.then (promiseResult) -> resultOfPromiseSearch = promiseResult
+
+            runs ->
+              expect(resultOfPromiseSearch).toBe('cancelled')
```

Comment:
On further consideration, this behavior might be hard to change because of the following:

```coffee
    cancellablePromise = new Promise (resolve, reject) ->
      onSuccess = ->
        resolve(null)
        return
      onFailure = ->
        resolve('cancelled')
        return
      searchPromise.then(onSuccess, onFailure)
```

I guess we want to say:
* If all the searches succeed, `cancellablePromise` resolves with `null`.
* If a search is cancelled, `cancellablePromise` resolves with `'cancelled'`.
* If a search fails, `cancellablePromise` rejects.

I think the tricky part is that it requires us to distinguish between a search that is cancelled vs. one that fails. This will probably require hoisting of resolve/reject functions...